Flake: TestPlugins_CatchStderr

I've found that this test is flaking out sometimes. The cause appears to be when the plugin fails but doesn't print to stderr. The PluginConnection code formats the error differently depending on if it captured anything to stderr or not.

https://github.com/getporter/porter/blob/59c4a56aabb6695ad970b60154a7316b58d79cdb/pkg/plugins/pluggable/connection.go#L158-L161

```
2022-06-29T15:21:33.5652721Z === RUN   TestPlugins_CatchStderr
2022-06-29T15:21:33.5653045Z === RUN   TestPlugins_CatchStderr/plugin_throws_an_error
2022-06-29T15:21:33.5653413Z     pluggable_integration_test.go:59: 
2022-06-29T15:21:33.5653859Z         	Error Trace:	pluggable_integration_test.go:59
2022-06-29T15:21:33.5654210Z         	Error:      	Error message not equal:
2022-06-29T15:21:33.5655272Z         	            	expected: "could not connect to the secrets.testplugin.vault plugin: Unrecognized remote plugin message: \n\nThis usually means that the plugin is either invalid or simply\nneeds to be recompiled to support the latest protocol.: i am just test plugin. i don't function\n"
2022-06-29T15:21:33.5656209Z         	            	actual  : "Unrecognized remote plugin message: \n\nThis usually means that the plugin is either invalid or simply\nneeds to be recompiled to support the latest protocol."
2022-06-29T15:21:33.5656791Z         	Test:       	TestPlugins_CatchStderr/plugin_throws_an_error
2022-06-29T15:21:33.5657399Z --- FAIL: TestPlugins_CatchStderr (0.76s)
2022-06-29T15:21:33.5657946Z     --- FAIL: TestPlugins_CatchStderr/plugin_throws_an_error (0.03s)
```

I'm not sure why sometimes that test plugin outputs to stderr and sometimes it doesn't? I may just be confused though. 😅

--- a/cmd/testplugin/main.go
+++ b/cmd/testplugin/main.go
@@ -3,8 +3,11 @@ package main
 import (
 	"fmt"
 	"os"
+	"time"
 )
 
 func main() {
 	fmt.Fprintln(os.Stderr, "i am just test plugin. i don't function")
+	time.Sleep(10 * time.Millisecond) // Hack to workaround the stderr message not being captured by the time the malformed stdout is processed.
+	fmt.Fprintln(os.Stdout, "i am a little teapot")
 }

--- a/magefile.go
+++ b/magefile.go
@@ -430,7 +430,7 @@ func chmodRecursive(name string, mode os.FileMode) error {
 
 // Run integration tests (slow).
 func TestIntegration() {
-	mg.Deps(tests.EnsureTestCluster, copySchema)
+	mg.Deps(tests.EnsureTestCluster, copySchema, BuildTestPlugin)
 
 	var run string
 	runTest := os.Getenv("PORTER_RUN_TEST")
@@ -442,8 +442,12 @@ func TestIntegration() {
 	if mg.Verbose() {
 		verbose = "-v"
 	}
-	must.RunV("go", "build", "-o", "bin/testplugin", "./cmd/testplugin")
+
 	must.Command("go", "test", verbose, "-timeout=30m", run, "-tags=integration", "./...").CollapseArgs().RunV()
+}
+
+func BuildTestPlugin() {
+	must.RunV("go", "build", "-o", "bin/testplugin", "./cmd/testplugin")
 }
 
 // Copy the locally built porter and exec binaries to PORTER_HOME

--- a/pkg/plugins/pluggable/connection.go
+++ b/pkg/plugins/pluggable/connection.go
@@ -156,9 +156,11 @@ func (c *PluginConnection) Start(ctx context.Context, pluginCfg io.Reader) error
 	span.Debug("Connecting to plugin", attribute.String("plugin-command", strings.Join(c.pluginCmd.Args, " ")))
 	rpcClient, err := c.client.Client(ctx)
 	if err != nil {
-		if stderr := errbuf.String(); stderr != "" {
-			err = fmt.Errorf("could not connect to the %s plugin: %w: %s", c.key, err, stderr)
+		pluginErr := errbuf.String()
+		if pluginErr != "" {
+			pluginErr = ": plugin stderr was " + pluginErr
 		}
+		err = fmt.Errorf("could not connect to the %s plugin%s: %w", c.key, pluginErr, err)
 		span.Error(err) // Emit the error before trying to close the connection
 		c.Close(ctx)
 		return err

--- a/tests/integration/pluggable_integration_test.go
+++ b/tests/integration/pluggable_integration_test.go
@@ -3,8 +3,10 @@
 package integration
 
 import (
+	"bytes"
 	"os/exec"
 	"path"
+	"path/filepath"
 	"testing"
 
 	"get.porter.sh/porter/pkg/config"
@@ -16,19 +18,31 @@ import (
 
 func TestPlugins_CatchStderr(t *testing.T) {
 	c := config.NewTestConfig(t)
+	defer c.Close()
 	ctx, _, _ := c.SetupIntegrationTest()
 
 	t.Run("plugin throws an error", func(t *testing.T) {
 		pluginsPath, _ := c.GetPluginsDir()
 		pluginName := "testplugin"
 
-		err := exec.Command("mkdir", "-p", path.Join(pluginsPath, pluginName)).Run()
+		pluginDir := path.Join(pluginsPath, pluginName)
+		err := exec.Command("mkdir", "-p", pluginDir).Run()
 		require.NoError(t, err, "could not create plugin dir")
 
 		// testplugin binary will be in bin. refer "test-integration" in Makefile
 		binDir := c.TestContext.FindBinDir()
-		err = exec.Command("cp", path.Join(binDir, pluginName), path.Join(pluginsPath, pluginName)).Run()
+		err = exec.Command("cp", path.Join(binDir, pluginName), pluginDir).Run()
 		require.NoError(t, err, "could not copy test binary")
+
+		// Verify that our test plugin will return a predictable error message when run
+		var pluginStderr bytes.Buffer
+		var pluginStdout bytes.Buffer
+		pluginCmd := exec.Command(filepath.Join(pluginDir, pluginName))
+		pluginCmd.Stdout = &pluginStdout
+		pluginCmd.Stderr = &pluginStderr
+		require.NoError(t, pluginCmd.Run(), "error running test plugin standalone")
+		require.Equal(t, "i am just test plugin. i don't function\n", pluginStderr.String(), "the test plugin isn't printing the expected message to stderr")
+		require.Equal(t, "i am a little teapot\n", pluginStdout.String(), "the test plugin isn't printing the expected message to stdout")
 
 		cfg := pluggable.PluginTypeConfig{
 			Interface: plugins.PluginInterface,
@@ -56,10 +70,10 @@ func TestPlugins_CatchStderr(t *testing.T) {
 		if conn != nil {
 			conn.Close(ctx)
 		}
-		require.EqualError(t, err, `could not connect to the secrets.testplugin.vault plugin: Unrecognized remote plugin message: 
+		require.EqualError(t, err, `could not connect to the secrets.testplugin.vault plugin: plugin stderr was i am just test plugin. i don't function
+: Unrecognized remote plugin message: i am a little teapot
 
 This usually means that the plugin is either invalid or simply
-needs to be recompiled to support the latest protocol.: i am just test plugin. i don't function
-`)
+needs to be recompiled to support the latest protocol.`)
 	})
 }

